src/library/autodj/trackselector.h

```diff
@@ -0,0 +1,94 @@
+#ifndef TRACKSELECTOR_H
+#define TRACKSELECTOR_H
+
+#include <QObject>
+
+#include "library/playlisttablemodel.h"
+#include "track/track.h"
+#include "util/class.h"
+
+
+class ITrackSelector
+{
+  public:
+    virtual ~ITrackSelector(){} // do not forget this
+
+    virtual const QString getName() = 0;
+
+    virtual const QString getId() = 0;
+
+    virtual TrackPointer getNextTrackFromQueue() = 0;
+
+
+};
+
+Q_DECLARE_INTERFACE(ITrackSelector, "ITrackSelector") // define this out of namespace scope
+
+
+/*!
+    \class TrackSelector
+    \brief Base class that is responsible for selecting the next AutoDJ tack
+*/
+
+class TrackSelector : public QObject, public ITrackSelector {
+    Q_OBJECT
+    Q_INTERFACES(ITrackSelector)
+  public:
+    TrackSelector(AutoDJProcessor* processor);
+
+  protected:
+    AutoDJProcessor* m_processor;
+    char* m_id;
```

Comment:
No need to store this twice. You can call the virtual getID() whenever you need it.

Reply:
You can pass AutoDJProcessor* as a parameter to each function invocation. This can be done by using a plain pointer, that must only be used during the invocation:

TrackPointer getNextTrackFromQueue(AutoDJProcessor* pAutoDjProcessor) ;

I don't see any reason why a TrackSelector should store reference to AutoDJProcessor between invocations.